Add SpanCategorizer component (#6747)
explosion/spaCy: website/docs/api/spancategorizer.md

---
title: SpanCategorizer

> tag: class

Reviewer: ```suggestion
tag: class,experimental
```

source: spacy/pipeline/spancat.py

> new: 3

Reviewer: ```suggestion
new: 3.1
```

---
title: SpanCategorizer
tag: class
source: spacy/pipeline/spancat.py
new: 3
teaser: 'Pipeline component for labeling potentially overlapping spans of text'
api_base_class: /api/pipe
api_string_name: spancat
api_trainable: true
---

A span categorizer consists of two parts: a [suggester function](#suggesters)
that proposes candidate spans, which may or may not overlap, and a labeler model

> that predicts one or more labels for each span.

Author: ```suggestion
that predicts zero or more labels for each candidate.
```

modifying them.

> #### Example
>
> ```python
> spancat = nlp.add_pipe("spancat")
> scores = spancat.predict([doc1, doc2])
> ```

| Name        | Description                                 |
| ----------- | ------------------------------------------- |
| `docs`      | The documents to predict. ~~Iterable[Doc]~~ |
| **RETURNS** | The model's prediction for each document.   |

## SpanCategorizer.set_annotations {#set_annotations tag="method"}

Modify a batch of [`Doc`](/api/doc) objects using pre-computed scores.

> #### Example
>
> ```python
> spancat = nlp.add_pipe("spancat")
> scores = spancat.predict(docs)
> spancat.set_annotations(docs, scores)
> ```

| Name     | Description                                               |
| -------- | --------------------------------------------------------- |
| `docs`   | The documents to modify. ~~Iterable[Doc]~~                |
| `scores` | The scores to set, produced by `SpanCategorizer.predict`. |

## SpanCategorizer.update {#update tag="method"}

Learn from a batch of [`Example`](/api/example) objects containing the
predictions and gold-standard annotations, and update the component's model.

> Delegates to [`predict`](/api/spancategorizer#predict) and
> [`get_loss`](/api/spancategorizer#get_loss).

Reviewer 2: ```suggestion
Delegates to [`predict`](/api/spancategorizer#predict),
[`get_loss`](/api/spancategorizer#get_loss) and 
[`set_annotations`](/api/spancategorizer#set_annotations).
```

> #### Example
>
> ```python
> spancat = nlp.add_pipe("spancat")
> optimizer = nlp.initialize()
> losses = spancat.update(examples, sgd=optimizer)
> ```

| Name              | Description                                                                                                                        |
| ----------------- | ---------------------------------------------------------------------------------------------------------------------------------- |
| `examples`        | A batch of [`Example`](/api/example) objects to learn from. ~~Iterable[Example]~~                                                  |
| _keyword-only_    |                                                                                                                                    |

> | `drop`            | The dropout rate. ~~float~~                                                                                                        |
> | `set_annotations` | Whether or not to update the `Example` objects with the predictions, delegating to [`set_annotations`](#set_annotations). ~~bool~~ |
> | `sgd`             | An optimizer. Will be created via [`create_optimizer`](#create_optimizer) if not set. ~~Optional[Optimizer]~~                      |

Reviewer 2: ```suggestion
| `drop`            | The dropout rate. ~~float~~                                                                                                        |
| `sgd`             | An optimizer. Will be created via [`create_optimizer`](#create_optimizer) if not set. ~~Optional[Optimizer]~~                      |
```
